src/condor_utils/file_transfer.cpp

```diff
@@ -4141,7 +4152,47 @@ FileTransfer::DoUpload(filesize_t *total_bytes_ptr, ReliSock *s)
 		}
 	}
 
+
+	//
+	// Remove file entries which result in duplicates at the destination.
+	// This is no longer necessary for correctness (because I changed
+	// FileTransferItem::operator < to group rather than sort), but it's
+	// still more efficient.  Some of these duplicates will have been
+	// generated internally, but not all.  It's more efficient to remove
+	// them all here (especially when the duplicate would only have been
+	// generated by expanding a directory).  We remove duplicate directory
+	// creation entries when we create the extra entries because for those,
+	// we want to preserve the earlier one, rather than the later one.
+	//
+	std::set<std::string> names;
+	for( auto iter = filelist.rbegin(); iter != filelist.rend(); ++iter ) {
```

Comment:
This is, worst case, O(N^2), right?  (`filelist.erase` has linear complexity based on the number of elements that have to shift in the vector).

Why not:
1.  Iterate once to come up with a list of duplicates.  Linear complexity.
2.  Invoke `remove_if` with a lambda that just hits the duplicates.
3.  Erase the remaining contents.

That keeps things linear in the end.

(NOTE: this conflicts with #258 which coalesces duplicates.  Will require some careful merging)

Reply:
As far as I can tell, the entries are permitted to be identical except for their position in the list.  That would normally mean recording the iterators of the matching elements, except that removing elements invalidates other iterators.  Even if we only need a subset of attributes to identify the entries in your proposed step 2, given how the short the list typically is, I'm considerably more worried about clear and simple code than memory usage, so I'd prefer to linearize the code by copying the elements to a new list without duplicates.

Reply:
There is `std::unique` ...

Reply:
... which won't work in this situation, sadly (because it requires the list be sorted according to the equality comparison).

Reply:
If we're going for simplicity, how about just keep the duplicates?

[Anyhow, I'm ambivalent on this one because #258 implements the linear sort and we'll have to fix it up in the merge of that PR.]

Reply:
I don't think it's appropriate to make all of users wait for duplicate transfers.  If the approving reviewer thinks that O(n^2) matters for as few elements as are typically in the file transfer lists, I'll convert to a linear algorithm; otherwise, if it happens during the merge of #258, that's fine too.